File: isingWorm.py
import sys
SAVEFRAMES = False
from math import tanh, pow
from indexClusters import *
from graphs import *
from plotGraph import *
from utils import *

def isAccepted(K, site0, site1, gitter):
    """Check if the link between site0 and site1 in gitter is accepted

    :K: Float - J/T
    :site0: 1x2 matrix - [i, j]
    :site1: 1x2 matrix - [i, j]
    :gitter: dictionary
    :returns: Boolean - True if accepted and False otherwise
    """

    # Convert sites to a tuples (tuples can be used as hash values but not arrays)
    site0 = tuple(site0)
    site1 = tuple(site1)

    linkWeight = getLinkWeight(site0, site1, gitter)

    # Probability of being accepted
    p = pow(tanh(K), 1 - linkWeight)

    if random.random() < p:
        return True
    else:
        return False

def saveFrame(obj, typeOfObj, seed, N, M, boundaryCondition, numberOfFrames):
    """ Save obj to ./data/{N}x{M}_{boundaryCondition}/{seed}/{typeOfObj}/frameNumber.pickle

    :obj: Something to save
    :typeOfObj: String - One of ["gitter", "correlation_function", "clusters"]
    :seed: Int
    :N: Int - Number of rows
    :M: Int - Number of columns
    :boundaryCondition: String
    :returns: None
    """
    if typeOfObj not in ["gitter", "correlation_function", "clusters"]:
        raise Exception(f"Type of object to save in saveFrame not supported. Recieved {typeOfObj}.")

    path = f"data/{N}x{M}_{boundaryCondition}/{seed}/{typeOfObj}/{numberOfFrames}"

    # Save the next frame
    saveObject(obj, path)

def updateCorrFunc(firstSite, nextSite, corrFunction):
    """Update the correlation function

    :corrFunction: dictionary
    :nextSite: 1x2 matrix
    :startSite: 1x2 matrix
    :returns: None
    """
    # Update correlation function on the fly:
    # add +1 to G(i-i0) for the open path from i0 to i
    addIfNotExists(firstSite[0] - nextSite[0], 1, corrFunction)

def updateLoopLengths(clusters):
    """Append the number of sites in clusters with index to loopLengths

    :loopLengths: 1xn matrix - list of loop lengths
    :index: Int - cluster index
    :clusters: dictionary - {index: listOfSitesInCluster, ...}
    :returns: None
    """
    loopLengths = []
    for index in clusters:
        loopLengths.append(len(clusters[index]))
    return loopLengths

def simulateWorm(corrFunction, K, N, M, gitter, seed, numberOfFrames):
    """Start a new worm and moves it until a loop is formed
    Chooses any neighbour from currentSite except previousSite to avoid moving 180 degrees.
    Updates the correlation function corrFunction after each step.

    :K: J/T
    :M: Int - Number of columns
    :N: Int - Number of rows
    :currentSite: 1x2 matrix
    :gitter: dictionary
    :previousSite: 1x2 matrix
    :returns: 1xn matrix - loop lengths
    """

    # Initialize starting site as some random [i, j] within the gitter
    firstSite = [random.randrange(0, N), random.randrange(0, M)]

    if DEBUG:
        print(f"First site: {firstSite}")

    # Get some random neighbour to form the first link
    currentSite = getRandomNeighbour(firstSite, None, gitter)
    switchLinkBetween(currentSite, firstSite, gitter)

    if DEBUG:
        print(f"Initial color link between {firstSite} and {currentSite}")

    # Track the previous site to avoid that the current turns 180 degrees
    previousSite = firstSite

    clusters = {}
    indexClusters(clusters, gitter)

    updateCorrFunc(firstSite, currentSite, corrFunction)

    loopFormed = False
    while not loopFormed:
        # Get potential next step (choose any neighbour exept previousSite)
        nextSite = getRandomNeighbour(site=currentSite, exceptSite=previousSite, graph=gitter)

        if isAccepted(K, currentSite, nextSite, gitter):

            if DEBUG:
                print(f"Accepted site {nextSite}")
                print(f"Coloring from {currentSite} to {nextSite}")

            # Flip the weight between currentSite and nextSite
            switchLinkBetween(currentSite, nextSite, gitter)

            previousSite = currentSite
            currentSite = nextSite

            if DEBUG:
                print(f"Updating indexing...")

            # Update indexing
            clusters = {}
            indexClusters(clusters, gitter)

            if DEBUG:
                print(f"The cluster is now: {clusters}")

            if nextSite == firstSite:
                # Found a new loop
                loopFormed = True

                if DEBUG:
                    print(f"Found a looop on index {indexOfLoop}")
                    print(f"Loop lengths before updating: {loopLengths}")

                loopLengths = updateLoopLengths(clusters)

                if DEBUG:
                    print(f"Loop lengths after updating: {loopLengths}")

            # Update the correlation function when new site is accepted
            updateCorrFunc(firstSite, nextSite, corrFunction)

            if SAVEFRAMES:
                # Save the frame
                saveFrame(corrFunction, "gitter", seed, N, M, boundaryCondition, numberOfFrames)
                saveFrame(corrFunction, "correlation_function", seed, N, M, boundaryCondition, numberOfFrames)
                saveFrame(clusters, "clusters", seed, N, M, boundaryCondition, numberOfFrames)
            numberOfFrames += 1

            # Plotting
            plotGraph(clusters, gitter)
            plotCorr(corrFunction, M)

    return loopLengths

def main(K, N, M, boundaryCondition):
    """Simulate ising worm algorithm

    :K: Float - J/T
    :N: Int - Number of rows
    :M: Int - Number of columns
    :boundaryCondition: String - Supported: ["periodic", "dirichlet"]
    :returns: dictionaries - gitter, correlation function
    """

    # Initialize the random number generator with current time as seed
    seed = random.randrange(sys.maxsize)

    # Self-eating loop (resulting in clusters = {})
    seed = 5540102676881230539
    # seed = 1592744071574553462

    # Multiple loops but only one gets indexed
    # seed = 595770392852380573

    # Produces a nice gif. loopLengths = [4, 4, 12, 4]
    # seed = 615885798301355417

    random.seed(seed)
    print(f"The seed is: {seed}")

    gitter = buildGraph(N, M, boundaryCondition)

    # Correlation function for x
    corrFunction = {}

    # How many frames have been played
    numberOfFrames = 0
    for _ in range(2):
        # Move this worm until it forms a loop
        loopLengths = simulateWorm(corrFunction, K, N, M, gitter, seed, numberOfFrames)

    if len(loopLengths) != 0:
        # Update the average loop length
        averageLoopLength = sum(loopLengths)/len(loopLengths)
    else:
        raise Exception(f"There were no loops. Loop lengths: {loopLengths}")

    if DEBUG:
        print(f"Loop lengths: {loopLengths}")
        print(f"Loop averages: {averageLoopLength}")

    return gitter, corrFunction, averageLoopLength, seed

if __name__ == '__main__':
    J = 0.5
    # Temperature
    T = 1

    K = J/T

    # Number of rows in gitter
    N = 10
    # Number of columns in gitter
    M = 10

    boundaryCondition = "dirichlet"
    # boundaryCondition = "periodic"

    # Run the simulation
    gitter, corrFunction, averageLoopLength, seed = main(K, N, M, boundaryCondition) 
    # Keep the window open when done
    plt.show()
    if DEBUG:
        print(f"The seed was: {seed}")

File: cpp/source/Ising/IsingWorm.cpp
#include "Graph.h"
#include "utils.h"
#include <iostream>
#include <string>
#include <unordered_map>

/**
* @brief: Set the absolute difference in x value between site0 and site1 as a key if it doesn't already exist.
*         Then add +1 to the corresponding value.
*
* @param: int site0
*       : int site1
*       : int length
*       : std::unordered_map<int, int> &correlation_func
*
* @return: void
*/
void UpdateCorrelationFunction(int site0, int site1, int length, std::unordered_map<int, int> &correlation_func) {
    // # add +1 to G(i-i0) for the open path from i0 to i
    // # NOTE: This has to be the absolute value,
    // #       otherwise it will be skewed toward the side with the largest number of sites.

	// std::cout << "Call to UpdateCorrelationFunction" << "\n";
	// std::cout << "Site input: " << site0 << ", " << site1 << "\n";

	// Get the x values for each site
	int x_0 = site0 % length;
	int x_1 = site1 % length;

	// std::cout << "x value for site " << site0 << " is " << x_0 << "\n";
	// std::cout << "x value for site " << site1 << " is " << x_1 << "\n";

	// Get the absolute number of the difference between the x values
	int key = (x_0 > x_1) ? x_0 - x_1 : x_1 - x_0;

	// std::cout << "Correlation function is as: " << "\n";
	// for (auto element : correlation_func) {
	// 	std::cout << element.first << ": "<< element.second << "\n";
	// }

	if (MapHasItem(key, correlation_func)) {
		// std::cout << "Adding +1 to old key: " << key << "\n";
		correlation_func[key]++;
	} else {
		// std::cout << "Adding +1 to new key: " << key << "\n";
		correlation_func[key] = 1;
	}
}

/**
* @brief: Check if the link between current_site and next_site in gitter is accepted
*
* @param: double K
*       : int current_site
*       : int next_site
*       : bool link_between
*       : bool random_num
*
* @return: bool
*/
bool IsAccepted(double K, bool link_between, long double &random_num) {
    // Probability of being accepted
    auto p = std::pow(std::tanh(K), 1 - link_between);

	if (random_num < p) {
		return 1;
	} else {
		return 0;
	}
}

/**
* @brief: Get the average loop length weighted with tanh(K)
*         \sum { l * tanh^l(K) }
*         ----------------------
*           \sum { tanh^l(K) }
*
* @param: std::unordered_map<int, int> &loop_lengths
*       : double const &K
*
* @return: double
*/
double GetAverageLoopLength(std::unordered_map<int, int> &loop_lengths, double const &K) {
	// TODO: Test this function

	//    \sum { l * tanh^l(K) }
	double sum_above = 0;

	//    \sum { tanh^l(K) }
	double sum_below = 0;

	for (auto& index_and_loop_length : loop_lengths) {
		double tanhK_to_l = std::pow(std::tanh(K), index_and_loop_length.second);
		sum_above += index_and_loop_length.second * tanhK_to_l;
		sum_below += tanhK_to_l;
	}
	// \sum { l * tanh^l(K) }
	// ----------------------
	//   \sum { tanh^l(K) }
	return sum_above / sum_below;
}

/**
* @brief: For each cluster in clusters, find the length of that loop
*
* @param: std::unordered_map<int, int> &loop_lengths
*       : std::unordered_map<int, std::vector<int> &clusters
*       : Graph &lattice
*
* @return: void
*/
void UpdateLoopLengths(std::unordered_map<int, int> &loop_lengths, std::unordered_map<int, std::vector<int>> &clusters, Graph &lattice) {
	//TODO: Test this function

	// std::cout << "Call to UpdateLoopLengths" << "\n";

	// Will be used to measure the length of all links from site
	std::vector<int> linked_neighbours;
	for (auto index_and_sites : clusters) {
		// Start each cluster with zero length
		int current_length = 0;

		// std::cout << "On index: " << index_and_sites.first << "\n";

		for (int site : index_and_sites.second) {
			// Find the number of links going through site
			// NOTE: This will be double counting since if
			//       site0 is a linked neighbour to site1
			//       site1 is a linked neighbour to site0.
			lattice.GetLinkedNeighbours(site, linked_neighbours);
			current_length += linked_neighbours.size();

			// std::cout << "Added site: " << site << "\n";
			// std::cout << "That has the neighbours: " << "\n";
			// for (auto n : linked_neighbours) {
			// 	std::cout << n << " ";
			// }
			// std::cout << "\nCurrent length is now: " << current_length << "\n";

			// Reinitialize the vector to be used on the next site
			linked_neighbours.clear();
		}

		// std::cout << "This cluster size is: " << current_length / 2 << "\n";

		// Divide by 2 to avoid double counting
		loop_lengths[index_and_sites.first] = current_length / 2;
	}
}

/**
* @brief: Run the simulation until at least one loop is formed and return the number of steps taken
*
* @param: Graph &lattice
*       : std::vector<int> correlation_func
*       : std::vector<int>> dimensions
*       : int length
*       : double K
*
* @return: long double
*/
long double IsingSimulation(Graph & lattice, double K) {
	// Get the first site for this simulation
	int first_site = lattice.GetRandomSite();

	// std::cout << "Got the first site: " << first_site << "\n";

	// Get some random neighbour to form the first link
	int current_site = lattice.GetRandomNeighbour(first_site);

	// std::cout << "Got the neighbour: " << current_site << "\n";

	// Form the first link
	lattice.SwitchLinkBetween(first_site, current_site);

	// Store the total number of accepted steps
	long double num_steps = 1.0;

	bool loop_formed = 0;
	while (!loop_formed) {
		int next_site = lattice.GetRandomNeighbour(current_site);

		// std::cout << "Next site is: " << next_site << "\n";

		auto rand_num = lattice.GetRandomNum();
		if (IsAccepted(K, lattice.GetLink(current_site, next_site), rand_num)) {
			num_steps++;

			// std::cout << "Got accepted!" << "\n";
			// std::cout << "Switch link between sites: " << current_site << " and " << next_site << "\n";

			// Flip the weight between currentSite and nextSite
			lattice.SwitchLinkBetween(current_site, next_site);
			current_site = next_site;
			// If we have found a loop
			if (next_site == first_site) {
				// std::cout << "Found a loop!" << "\n==========================\n\n";
				loop_formed = 1;

			}
		}
	}
	return num_steps;
}

/**
* @brief: Run warm_up_runs number of updates and throw them away
*
* @param: int warm_up_runs
*       : Graph& lattice
*
* @return: void
*/
void WarmUp(int warm_up_runs, Graph& lattice, double K) {
	std::cout << "Ising warmup called" << "\n";
	for (int i = 0; i < warm_up_runs; ++i) {
		IsingSimulation(lattice, K);
	}
}
